swag rejects valid MIME Types for @Produce, panic

**Describe the bug**
Swag limits the MIME types allowed for APIs without good reason to do so.
Here are four MIME types which I want to use which swag disallows, panicing when I use them:

* `application/xhtml+xml`
* `application/xml`
* `application/health+json`
* `image/svg+xml`

**To Reproduce**
Steps to reproduce the behavior:
1. Create an endpoint that produces a valid MIME type not in the documentation of `swag`.
2. Run `swag init`

**Expected behavior**
All valid MIME types are accepted. There are no restrictions to what APIs can use. APIs are free to choose.

** Actual Behavior**
`swag` panics.

```
+ /home/christian.hujer/go/bin//swag init
2019/02/20 16:16:44 Generate swagger docs....
2019/02/20 16:16:44 Generate general API Info
2019/02/20 16:16:44 ParseComment panic in file health/health.go :application/health+json produce type can't accepted
panic: ParseComment panic in file health/health.go :application/health+json produce type can't accepted

goroutine 1 [running]:
log.Panicf(0x8f0078, 0x22, 0xc4201b3510, 0x2, 0x2)
        /usr/lib/go-1.10/src/log/log.go:333 +0xda
github.com/swaggo/swag.(*Parser).ParseRouterAPIInfo(0xc4204b7630, 0xc4203f7240, 0x10, 0xc420502b00)
        /home/christian.hujer/go/src/github.com/swaggo/swag/parser.go:347 +0x256
github.com/swaggo/swag.(*Parser).ParseAPI(0xc4204b7630, 0x8e2247, 0x2, 0x8e3479, 0x7, 0x8e7ed3, 0x13)
        /home/christian.hujer/go/src/github.com/swaggo/swag/parser.go:96 +0x242
github.com/swaggo/swag/gen.(*Gen).Build(0xc42048f988, 0x8e2247, 0x2, 0x8e3479, 0x7, 0x8e5a3e, 0xe, 0x8e3e9a, 0x9, 0x0, ...)
        /home/christian.hujer/go/src/github.com/swaggo/swag/gen/gen.go:30 +0x2c6
main.main.func1(0xc4200fe2c0, 0xc4202efb00, 0xc4200fe2c0)
        /home/christian.hujer/go/src/github.com/swaggo/swag/cmd/swag/main.go:34 +0x194
github.com/urfave/cli.HandleAction(0x843500, 0x8ffde0, 0xc4200fe2c0, 0x0, 0xc420143020)
        /home/christian.hujer/go/src/github.com/urfave/cli/app.go:502 +0xbe
github.com/urfave/cli.Command.Run(0x8e285f, 0x4, 0x0, 0x0, 0xc42040f210, 0x1, 0x1, 0x8e5ae6, 0xe, 0x0, ...)
        /home/christian.hujer/go/src/github.com/urfave/cli/command.go:165 +0x47d
github.com/urfave/cli.(*App).Run(0xc420074000, 0xc4200ba000, 0x2, 0x2, 0x0, 0x0)
        /home/christian.hujer/go/src/github.com/urfave/cli/app.go:259 +0x6e8
main.main()
        /home/christian.hujer/go/src/github.com/swaggo/swag/cmd/swag/main.go:61 +0x4e2
```

**Your swag version**
```
~/go/bin/swag --version
swag version v1.4.1
```

**Desktop (please complete the following information):**
- OS: Kubuntu 18.04.2 LTS
- Go version: go1.10.4 linux/amd64

**Additional context**
I've seen that `swag` has already been patched several times to accomodate more MIME Types.
It should not be necessary to patch `swag` for new MIME Types.
`swag` should not panic on MIME Types it does not know, it should simply pass them through.
If `swag` shall help with avoiding accidents with MIME Types, such as wrong MIME Types, it should issue a warning for unknown MIME types instead of panicing. Also, in that case, `swag` should not maintain authority over the list of known MIME types but accept IANA as authority, refering to the official list of registered MIME types at https://www.iana.org/assignments/media-types

--- a/README.md
+++ b/README.md
@@ -371,20 +371,23 @@ OPTIONS:
 
 ## Mime Types
 
-| Mime Type                         | annotation                                                |
-|-----------------------------------|-----------------------------------------------------------|
-| application/json                  | application/json, json                                    |
-| text/xml                          | text/xml, xml                                             |
-| text/plain                        | text/plain, plain                                         |
-| html                              | text/html, html                                           |
-| multipart/form-data               | multipart/form-data, mpfd                                 |
-| application/x-www-form-urlencoded | application/x-www-form-urlencoded, x-www-form-urlencoded  |
-| application/vnd.api+json          | application/vnd.api+json, json-api                        |
-| application/x-json-stream         | application/x-json-stream, json-stream                    |
-| application/octet-stream          | application/octet-stream, octet-stream                    |
-| image/png                         | image/png, png                                            |
-| image/jpeg                        | image/jpeg, jpeg                                          |
-| image/gif                         | image/gif, gif                                            |
+`swag` accepts all MIME Types which are in the correct format, that is, match `*/*`.
+Besides that, `swag` also accepts aliases for some MIME Types as follows:
+
+| Alias                 | MIME Type                         |
+|-----------------------|-----------------------------------|
+| json                  | application/json                  |
+| xml                   | text/xml                          |
+| plain                 | text/plain                        |
+| html                  | text/html                         |
+| mpfd                  | multipart/form-data               |
+| x-www-form-urlencoded | application/x-www-form-urlencoded |
+| json-api              | application/vnd.api+json          |
+| json-stream           | application/x-json-stream         |
+| octet-stream          | application/octet-stream          |
+| png                   | image/png                         |
+| jpeg                  | image/jpeg                        |
+| gif                   | image/gif                         |
 
 ## Security
 

--- a/operation.go
+++ b/operation.go
@@ -30,6 +30,23 @@ type Operation struct {
 // Regular expression for comment with response
 const responseCommentPattern = `([\d]+)[\s]+([\w\{\}]+)[\s]+([\w\-\.\/]+)[^"]*(.*)?`
 
+var mimeTypeAliases = map[string]string{
+	"json":                  "application/json",
+	"xml":                   "text/xml",
+	"plain":                 "text/plain",
+	"html":                  "text/html",
+	"mpfd":                  "multipart/form-data",
+	"x-www-form-urlencoded": "application/x-www-form-urlencoded",
+	"json-api":              "application/vnd.api+json",
+	"json-stream":           "application/x-json-stream",
+	"octet-stream":          "application/octet-stream",
+	"png":                   "image/png",
+	"jpeg":                  "image/jpeg",
+	"gif":                   "image/gif",
+}
+
+var mimeTypePattern = regexp.MustCompile("^[^/]+/[^/]+$")
+
 // NewOperation creates a new Operation with default properties.
 // map[int]Response
 func NewOperation() *Operation {
@@ -328,71 +345,26 @@ func (operation *Operation) ParseTagsComment(commentLine string) {
 
 // ParseAcceptComment parses comment for given `accept` comment string.
 func (operation *Operation) ParseAcceptComment(commentLine string) error {
-	accepts := strings.Split(commentLine, ",")
-	for _, a := range accepts {
-		switch a {
-		case "json", "application/json":
-			operation.Consumes = append(operation.Consumes, "application/json")
-		case "xml", "text/xml":
-			operation.Consumes = append(operation.Consumes, "text/xml")
-		case "plain", "text/plain":
-			operation.Consumes = append(operation.Consumes, "text/plain")
-		case "html", "text/html":
-			operation.Consumes = append(operation.Consumes, "text/html")
-		case "mpfd", "multipart/form-data":
-			operation.Consumes = append(operation.Consumes, "multipart/form-data")
-		case "x-www-form-urlencoded", "application/x-www-form-urlencoded":
-			operation.Consumes = append(operation.Consumes, "application/x-www-form-urlencoded")
-		case "json-api", "application/vnd.api+json":
-			operation.Consumes = append(operation.Consumes, "application/vnd.api+json")
-		case "json-stream", "application/x-json-stream":
-			operation.Consumes = append(operation.Consumes, "application/x-json-stream")
-		case "octet-stream", "application/octet-stream":
-			operation.Consumes = append(operation.Consumes, "application/octet-stream")
-		case "png", "image/png":
-			operation.Consumes = append(operation.Consumes, "image/png")
-		case "jpeg", "image/jpeg":
-			operation.Consumes = append(operation.Consumes, "image/jpeg")
-		case "gif", "image/gif":
-			operation.Consumes = append(operation.Consumes, "image/gif")
-		default:
-			return fmt.Errorf("%v accept type can't accepted", a)
-		}
-	}
-	return nil
+	return parseMimeTypeList(commentLine, &operation.Consumes, "%v accept type can't be accepted")
 }
 
-// ParseProduceComment parses comment for gived `produce` comment string.
+// ParseProduceComment parses comment for given `produce` comment string.
 func (operation *Operation) ParseProduceComment(commentLine string) error {
-	produces := strings.Split(commentLine, ",")
-	for _, a := range produces {
-		switch a {
-		case "json", "application/json":
-			operation.Produces = append(operation.Produces, "application/json")
-		case "xml", "text/xml":
-			operation.Produces = append(operation.Produces, "text/xml")
-		case "plain", "text/plain":
-			operation.Produces = append(operation.Produces, "text/plain")
-		case "html", "text/html":
-			operation.Produces = append(operation.Produces, "text/html")
-		case "mpfd", "multipart/form-data":
-			operation.Produces = append(operation.Produces, "multipart/form-data")
-		case "x-www-form-urlencoded", "application/x-www-form-urlencoded":
-			operation.Produces = append(operation.Produces, "application/x-www-form-urlencoded")
-		case "json-api", "application/vnd.api+json":
-			operation.Produces = append(operation.Produces, "application/vnd.api+json")
-		case "json-stream", "application/x-json-stream":
-			operation.Produces = append(operation.Produces, "application/x-json-stream")
-		case "octet-stream", "application/octet-stream":
-			operation.Produces = append(operation.Produces, "application/octet-stream")
-		case "png", "image/png":
-			operation.Produces = append(operation.Produces, "image/png")
-		case "jpeg", "image/jpeg":
-			operation.Produces = append(operation.Produces, "image/jpeg")
-		case "gif", "image/gif":
-			operation.Produces = append(operation.Produces, "image/gif")
-		default:
-			return fmt.Errorf("%v produce type can't accepted", a)
+	return parseMimeTypeList(commentLine, &operation.Produces, "%v produce type can't be accepted")
+}
+
+// parseMimeTypeList parses a list of MIME Types for a comment like
+// `produce` (`Content-Type:` response header) or
+// `accept` (`Accept:` request header)
+func parseMimeTypeList(mimeTypeList string, typeList *[]string, format string) error {
+	mimeTypes := strings.Split(mimeTypeList, ",")
+	for _, typeName := range mimeTypes {
+		if mimeTypePattern.MatchString(typeName) {
+			*typeList = append(*typeList, typeName)
+		} else if aliasMimeType, ok := mimeTypeAliases[typeName]; ok {
+			*typeList = append(*typeList, aliasMimeType)
+		} else {
+			return fmt.Errorf(format, typeName)
 		}
 	}
 	return nil

--- a/operation_test.go
+++ b/operation_test.go
@@ -46,10 +46,12 @@ func TestParseAcceptComment(t *testing.T) {
 		"application/octet-stream",
 		"image/png",
 		"image/jpeg",
-		"image/gif"
+		"image/gif",
+		"application/xhtml+xml",
+		"application/health+json"
     ]
 }`
-	comment := `/@Accept json,xml,plain,html,mpfd,x-www-form-urlencoded,json-api,json-stream,octet-stream,png,jpeg,gif`
+	comment := `/@Accept json,xml,plain,html,mpfd,x-www-form-urlencoded,json-api,json-stream,octet-stream,png,jpeg,gif,application/xhtml+xml,application/health+json`
 	operation := NewOperation()
 	err := operation.ParseComment(comment, nil)
 	assert.NoError(t, err)
@@ -79,10 +81,11 @@ func TestParseProduceComment(t *testing.T) {
 		"application/octet-stream",
 		"image/png",
 		"image/jpeg",
-		"image/gif"
+		"image/gif",
+		"application/health+json"
     ]
 }`
-	comment := `/@Produce json,xml,plain,html,mpfd,x-www-form-urlencoded,json-api,json-stream,octet-stream,png,jpeg,gif`
+	comment := `/@Produce json,xml,plain,html,mpfd,x-www-form-urlencoded,json-api,json-stream,octet-stream,png,jpeg,gif,application/health+json`
 	operation := new(Operation)
 	operation.ParseComment(comment, nil)
 	b, _ := json.MarshalIndent(operation, "", "    ")

